frame/uniques/src/types.rs

```diff
@@ -0,0 +1,117 @@
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Various basic types for use in the assets pallet.
+
+use super::*;
+
+pub(super) type DepositBalanceOf<T, I = ()> =
+	<<T as Config<I>>::Currency as Currency<<T as SystemConfig>::AccountId>>::Balance;
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug)]
+pub struct ClassDetails<
+	AccountId,
+	DepositBalance,
+> {
+	/// Can change `owner`, `issuer`, `freezer` and `admin` accounts.
+	pub(super) owner: AccountId,
+	/// Can mint tokens.
+	pub(super) issuer: AccountId,
+	/// Can thaw tokens, force transfers and burn tokens from any account.
+	pub(super) admin: AccountId,
+	/// Can freeze tokens.
+	pub(super) freezer: AccountId,
+	/// The balance deposited for this asset. This pays for the data stored here.
+	pub(super) deposit: DepositBalance,
+	/// The total balance deposited for the all metadata associated with this asset (both class and
+	/// instances).
+	pub(super) metadata_deposit: DepositBalance,
+	/// If `true`, then no deposit is needed to hold instances of this class.
+	pub(super) free_holding: bool,
+	/// The total number of outstanding unique instances of this asset class.
+	pub(super) instances: u32,
+	/// The number of instances of this asset class with a zero deposit.
+	pub(super) free_holds: u32,
+	/// Whether the asset is frozen for non-admin transfers.
+	pub(super) is_frozen: bool,
+}
+
+/// Witness data for the destroy transactions.
+#[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug)]
+pub struct DestroyWitness {
+	/// The number of instances of this class of asset.
+	#[codec(compact)]
+	pub(super) instances: u32,
+	/// The number of instances of this asset class with a zero deposit.
+	#[codec(compact)]
+	pub(super) free_holds: u32,
+}
+
+impl<AccountId, DepositBalance> ClassDetails<AccountId, DepositBalance> {
+	pub fn destroy_witness(&self) -> DestroyWitness {
+		DestroyWitness {
+			instances: self.instances,
+			free_holds: self.free_holds,
+		}
+	}
+}
+
+/// Information concerning the ownership of a single unique asset.
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, Default)]
+pub struct InstanceDetails<AccountId, DepositBalance> {
+	/// The owner of this asset.
+	pub(super) owner: AccountId,
+	/// The approved transferrer of this asset, if one is set.
+	pub(super) approved: Option<AccountId>,
```

Comment:
An API missing from this pallet which exists in ERC-721 is `set_approval_for_all`, which allows a user to easily delegate all of their uniques to another person, versus approving each unique one by one.

Reply:
I'm ok with it being introduced in a later PR.